rfcs/rfc-HAARG-0001.md

```diff
@@ -0,0 +1,180 @@
+# Map with different topic variable
+
+## Preamble
+
+    Author:  Graham Knop <haarg@haarg.org>
+    Sponsor:
+    ID: HAARG-0001
+    Status:  Draft
+
+## Abstract
+
+Allow `map` and `grep` to be given a different variable to be used as the topic
+variable, rather than using `$_`. Also allow multiple variables to be given to
+do n-at-a-time iteration.
+
+## Motivation
+
+Traditionally, `map` and `grep` loop over a list, aliasing `$_` to each list
+entry. While this can be convenient in many cases, it gets awkward when nested
+maps are needed. You need to manually save the `$_` value in another variable.
+
+```perl
+my %hash = (
+  foo => [ 1, 2, 3, 4 ],
+  bar => [ 5, 6, 7, 8 ],
+);
+
+my @out = map {
+  my $key = $_;
+  map { "$key: $_" } $hash{$key}->@*;
+} keys %hash;
+```
+
+Using `$_` can also be dangerous if you need to call code not under your
+direct control, due to it being the implicit target of operations like
+`readline`.
+
+It would be more convenient if you could use a different variable for the
+topic, similar to what `for` allows.
+
+```perl
+my @out = map my $key {
+  map my $i { "$key: $i" } $hash{$key}->@*;
+} keys %hash;
+```
+
+A natural extension of this syntax would be to allow n-at-a-time iteration, as
+`for` can do on perl 5.36+.
+
+```perl
+my @out = map my ($key, $val) {
+  map my $i { "$key: $i" } $val->@*;
+} %hash;
+```
+
+All of this would apply to `grep` as well.
```

Comment:
I agree with both feelings. I would consider the "undef" placeholders to be more of the result of needing something where there's nothing, rather than being treated as actual undef values; thus they should have no effect on the ultimate result when reasonable.

Reply:
I would say that the scalar value should be the size of the list it would return.

The list being returned should be aliases of the incoming values. That is how grep works now. So while missing values should be undef during the iteration, they should be missing in the return value.

I'll write up some additions to cover this.